signature/openpgp/openpgp.go

```diff
@@ -0,0 +1,86 @@
+package openpgp
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/containers/image/types"
+
+	"golang.org/x/crypto/openpgp"
+)
+
+type context struct {
+	keyring openpgp.EntityList
+}
+
+type OpenPGPMechanism struct {
+	ctx *context
+}
+
+func NewOpenPGPSigningMechanism() (types.SigningMechanism, error) {
+	return OpenPGPMechanism{ctx: &context{
+		keyring: openpgp.EntityList{},
+	}}, nil
+}
+
+func (m OpenPGPMechanism) ImportKeysFromBytes(blob []byte) ([]string, error) {
+	keyring, err := openpgp.ReadArmoredKeyRing(bytes.NewReader(blob))
+	if err != nil {
+		return nil, err
+	}
+	for _, entity := range keyring {
+		if entity.PrimaryKey == nil {
+			continue
+		}
+		m.ctx.keyring = append(m.ctx.keyring, entity)
+	}
+	if len(m.ctx.keyring) == 0 {
+		return nil, errors.New("no public keys found")
+	}
+	keyIdentities := []string{}
+	for _, entity := range m.ctx.keyring {
+		// Uppercase the fingerprint to be compatible with gpgme
+		keyIdentities = append(keyIdentities, strings.ToUpper(fmt.Sprintf("%x", entity.PrimaryKey.Fingerprint)))
+	}
+	return keyIdentities, nil
+}
+
+func (m OpenPGPMechanism) Sign(input []byte, keyIdentity string) ([]byte, error) {
+	return nil, errors.New("signing not implemented")
+}
+
+func (m OpenPGPMechanism) Verify(unverifiedSignature []byte) (contents []byte, keyIdentity string, err error) {
+	if len(m.ctx.keyring) == 0 {
+		return nil, "", errors.New("no public keys imported")
```

Comment:
Is this explicit check necessary? Admittedly it results in a nice error message, but it seems to me that “public key with fingerprint XXX not found” should happen anyway?!